File: 56. Merge Intervals.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */

struct cmp{
    bool operator() (Interval a, Interval b){
        return a.start > b.start || (a.start == b.start && a.end > b.end);
    }
};

class Solution {
public:
    // priority queue: O(NlogN)time & space, 68.53%, ok
    vector<Interval> merge(vector<Interval>& intervals) {
        vector<Interval> ret;
        if(intervals.empty()) return ret;
        priority_queue<Interval, vector<Interval>, cmp> pq;
        for(auto i: intervals) pq.push(i);

        int head = pq.top().start, tail = pq.top().end;
        pq.pop();
        while(pq.size()){
            int h = pq.top().start, t = pq.top().end;
            pq.pop();
            // bug: tail may be greater than t
            if(h <= tail) tail = max(t, tail);
            else{
                ret.push_back(Interval(head, tail));
                head = h, tail = t;
            }
        }
        ret.push_back(Interval(head, tail));
        return ret;
    }

    // sort with lambda: O(NlogN)time & space, 84.42%, ok
    vector<Interval> merge(vector<Interval>& intervals) {
        vector<Interval> ret;
        if(intervals.empty()) return ret;
        // learn lambda !!!!
        sort(intervals.begin(), intervals.end(), [](Interval a, Interval b) {return a.start < b.start;});
        int head = intervals[0].start, tail = intervals[0].end;
        for(int i = 1; i < intervals.size(); i++){
            int s = intervals[i].start, e = intervals[i].end;
            // bug: tail may be greater than t
            if(s <= tail) tail = max(e, tail);
            else{
                ret.push_back(Interval(head, tail));
                head = s, tail = e;
            }
        }
        ret.push_back(Interval(head, tail));
        return ret;
    }
};

File: 56. Merge Intervals.py
# Definition for an interval.
# class Interval(object):
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution(object):
    
    # sort: O(NlogN) time, O(1) space
    def merge(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: List[Interval]
        """
        if len(intervals) == 0:
            return []
        intervals = sorted(intervals, key = lambda x: (x.start, x.end))
        start, end, ret = intervals[0].start, intervals[0].end, []
        for i in xrange(1, len(intervals)):
            if end >= intervals[i].start:
                # end = intervals[i].end   # bug!
                end = max(intervals[i].end, end)
            else:
                ret.append(Interval(start, end))
                start, end = intervals[i].start, intervals[i].end
        ret.append(Interval(start, end))
        return ret
    
    # sort, more elegant: O(NlogN) time, no space
    def merge(self, intervals):
        ret = []
        intervals = sorted(intervals, key = lambda x: x.start)
        for elem in intervals:
            if ret and elem.start <= ret[-1].end:
                ret[-1].end = max(elem.end, ret[-1].end)
            else:
                ret.append(elem)
        return ret
            
        